### PR TITLE
ObjectCleaner should continue cleaning despite exceptions

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
ObjectCleaner inovkes a Runnable which may execute user code (FastThreadLocal#onRemoval) and therefore exceptions maybe thrown. If an exception is thrown the cleanup thread will exit prematurely and we may never finish cleaning up which will result in leaks.

Modifications:
- ObjectCleaner should suppress exceptions and continue cleaning

Result:
ObjectCleaner will reliably clean despite exceptions being thrown.

### CODE REVIEW

**Comment by Scottmitch** [comment]
the first commit is part of https://github.com/netty/netty/pull/7597

File: common/src/main/java/io/netty/util/internal/ObjectCleaner.java [hunk 1]
```diff
@@ -41,17 +47,21 @@ public void run() {
...
                     } catch (InterruptedException ex) {
                         // Just consume and move on
                         interrupted = true;
+                        continue;
+                    }
+                    if (reference != null) {
+                        try {
+                            reference.cleanup();
+                        } catch (Throwable ignored) {
+                            // ignore exceptions, and don't log in case the logger throws an exception.
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/ObjectCleaner.java]: I think we should try to log and just guard the logger itself. So there will be at least a chance the user see the exception

Comment (Scottmitch) [reply common/src/main/java/io/netty/util/internal/ObjectCleaner.java]: We can log but here are potential motivation to not log:
- the JDK also does the same:
https://github.com/netroby/jdk9-dev/blob/master/jdk/src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java#L153
- the logger may also block or have other side effects.

WDYT?

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (031bad60dcb70ea52697b09a2c17c62894b92c2b) and 4.0 (2fd6cb0a0f9ddaec03ef4837c06273d845dbb1fb)

**Comment by johnou** [comment]
@normanmaurer shouldn't we break out of the never ending loop after marking the thread interrupted?

**Comment by johnou** [comment]
Nevermind, it is outside of the for loop 🍺

**Pull request closed without merging**